### ISSUE TITLE
Extract conda-lock workflow into standalone GitHub Action?

### ISSUE THREAD

**Issue by weiji14** [CryoInTheCloud/hub-image#18]
Continuing discussion on whether to move some of the `conda-lock` related automation into a standalone GitHub Actions Continuous Integration workflow that can be shared with https://github.com/pangeo-data/pangeo-docker-images or other people.

> @weiji14 how do you feel about extracting the conda-lock stuff from here into an action by itself? Then we can use *that* action in pangeo-docker-images, thus removing duplicate work...

_Originally posted by @yuvipanda in https://github.com/CryoInTheCloud/hub-image/issues/13#issuecomment-1331657296_

To be clear, do we mean:
- [ ] The `/condalock` command added in #5, which I think pangeo-docker-images has a version of already maintained by @scottyhq?
- [ ] The 'installing from conda-lock.yml' script being worked on at #14? A quick search led to https://github.com/conda-incubator/conda-lock/issues/245#issuecomment-1263669546 and apparently there might be something re-usable already at https://github.com/mamba-org/provision-with-micromamba#example-with-customization?

**Comment by scottyhq** [comment]
hi @weiji14 i like the 2nd option of always installing from a lockfile. the `/condalock` command to be honest was a somewhat experimental thing that we tried early in learning github actions! the goal was to allow anyone to add packages (even from forks), but run the locking behind the scenes.

**Comment by weiji14** [comment]
> i like the 2nd option of always installing from a lockfile

Yeah, I agree that lockfiles are good for reproducible environments. I know pangeo-docker-images is currently installing from [explicit lockfiles](https://conda-incubator.github.io/conda-lock/output/#explicit-lockfile) (`conda-linux-64.lock`) at https://github.com/pangeo-data/pangeo-docker-images/blob/2e6d92f01e959ad2a5cc224e5090023dab81d828/base-image/Dockerfile#L155-L156, but the [unified lockfile (`conda-lock.yml`)](https://conda-incubator.github.io/conda-lock/output/#unified-lockfile) is something I'm keen to push on since it handles conda/pip dependencies, and is cross-platform Linux/macOS/Windows.

What are your thoughts @scottyhq from a maintenance perspective? This issue is really to discuss where best to put the 'install-from-lockfiles' piece of code in the context of docker images. Should this be added to `repo2docker`? I.e. this issue at https://github.com/jupyterhub/repo2docker/issues/1157? Or somewhere else?

And yes, the `/condalock` slash command is great :smile:

**Comment by scottyhq** [comment]
Good questions!

The goal is to have a reproducible environment installed either in CI or locally and right now you either have to 1. render a bunch of files with platform-specific locked packages or 2. have people install conda-lock and install the environment from there. (or 3. use Docker everywhere)

Short term, I'm supportive of moving to embracing a single conda-lock.yml and consequently micromamba or conda-lock in the CI workflow. 

I don't always want to use repo2docker because it can be helpful to have more control over the Dockerfile. But ultimately, a  way to contribute upstream and have the widest impact might be to help standardize the `conda-lock.yml` format. I don't know how much the people working on conda, mamba, and conda-lock are cross-communicating currently, but ideally the `conda-lock.yml` would be a standard recognized by different tools. Most people use `conda` or `mamba` locally and unfortunately you can't `mamba create -f conda-lock.yml` https://github.com/mamba-org/mamba/issues/1884...

**Comment by yuvipanda** [comment]
I'd love for the CondaLock.yml thingy to be packaged up and moved into a default, *especially* in a form that can also handle pip dependencies. Additionally, I think it should also autorun and not require a chatops command. If we can make that work, I'd love to incorporate that into https://github.com/2i2c-org/hub-user-image-template/ :)

**Comment by yuvipanda** [comment]
Coming back after the holidays, @weiji14 do you think you would be able to work on this? :)

**Comment by weiji14** [comment]
> Coming back after the holidays, @weiji14 do you think you would be able to work on this? :)

I'm currently back in Malaysia visiting family, and won't have much capacity until mid Feb at least. You can start working on some pieces firat if you're interested!

**Comment by weiji14** [comment]
> But ultimately, a way to contribute upstream and have the widest impact might be to help standardize the `conda-lock.yml` format. I don't know how much the people working on conda, mamba, and conda-lock are cross-communicating currently, but ideally the `conda-lock.yml` would be a standard recognized by different tools.

Just noticed that `conda-lock` graduated from conda-incubator to the main conda organisation at https://github.com/conda/conda-lock! Hopefully this means `conda-lock.yml` uptake will increase this year :smile:

My thought is that we should wait for a `conda-lock` 1.5.0 release (see https://github.com/conda/conda-lock/issues/336) for them to iron out any issues with the organization migration, and then someone could start working on the `/condalock` GitHub Action/App. There's also a big issue with `conda-lock` and `conda=22.11.1` at https://github.com/conda/conda-lock/issues/292 which I hit into recently, and I kinda want to see that resolved before we get working on anything serious.

Since this `/condalock` GitHub Action/App isn't cryo-specific, what organization do you think we should host it under? I see @scottyhq opened a parallel issue at https://github.com/pangeo-data/pangeo-docker-images/issues/414, so should it go under the `pangeo-data` org, or somewhere else?
